### PR TITLE
Loki: Add case insensitive line contains operation

### LINKED ISSUE #57051: Loki: Add case insensitive line contains operation
### Summary

With a new `Line contains case insensitive`  operation, we would significantly simplified and sped up a way, how to create a query that searches for text inside a log line ignoring the casing (uppercase/lowercase letters).

### Problem statement

For our users it is important to have a way create a query with case insensitive `Line contains` filter. Currently, it is possible to do it, but you need to manually add `(?i)` operator in `Line contains regex match` which requires LogQL knowledge. 

### Scope

Implementation (coding task)

### Solution

We should simplify how to create case insensitive line filter in Loki query builder. As we already have custom operations in query builder (e.g. `No pipeline errors`), adding `Line contains case insensitive`  would be a small change with big impact for some of our users, as they could use this operation instead of figuring out the correct regex pattern. 

Within this feature request, a new "Operation" should be added to the Loki Query Builder in this file: https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/querybuilder/operations.ts#L219

That operation should be called something along the lines of `Line contains case insensitive`. The operation should take the value of an input field and translate that into a LogQL RegEx operation, which is preceded with a `(?i)` and followed by the value.

Analogue to other line filters, this issue should also add a "Line does not contain case insensitive" operation, which will result in a `!~ "(?i)VALUE"` operation.

### PR DESCRIPTION
**Pull request by itsgareth**
**What is this feature?**
this PR adds a new operation in the loki query builder (`case insensitive line contains`)

**Why do we need this feature?**
...

**Who is this feature for?**
...

**Which issue(s) does this PR fix?**:

Fixes #57051

**Special notes for your reviewer**:
...

### CODE REVIEW

File: public/app/plugins/datasource/loki/querybuilder/operations.ts [hunk 1]
```diff
@@ -238,6 +238,28 @@ export function getOperationDefinitions(): QueryBuilderOperationDef[] {
       addOperationHandler: addLokiOperation,
       explainHandler: (op) => `Return log lines that contain string \`${op.params[0]}\`.`,
     },
+    {
+      id: LokiOperationId.LineContainsInsensitive,
+      name: 'Line contains (insensitive)',
```
Comment (matyax) [inline public/app/plugins/datasource/loki/querybuilder/operations.ts:243]:
```suggestion
      name: 'Line contains (case insensitive)',
```

File: public/app/plugins/datasource/loki/querybuilder/operations.ts [hunk 2]
```diff
@@ -260,6 +282,28 @@ export function getOperationDefinitions(): QueryBuilderOperationDef[] {
       addOperationHandler: addLokiOperation,
       explainHandler: (op) => `Return log lines that does not contain string \`${op.params[0]}\`.`,
     },
+    {
+      id: LokiOperationId.LineContainsNotInsensitive,
+      name: 'Line does not contain (insensitive)',
```
Comment (matyax) [inline public/app/plugins/datasource/loki/querybuilder/operations.ts:287]:
```suggestion
      name: 'Line does not contain (case insensitive)',
```

File: public/app/plugins/datasource/loki/querybuilder/operationUtils.ts [hunk 3]
```diff
@@ -257,6 +257,12 @@ export function addNestedQueryHandler(def: QueryBuilderOperationDef, query: Loki
...
   return function lineFilterRenderer(model: QueryBuilderOperation, def: QueryBuilderOperationDef, innerExpr: string) {
+    if (
+      model.id === LokiOperationId.LineContainsInsensitive ||
+      model.id === LokiOperationId.LineContainsNotInsensitive
+    ) {
+      return `${innerExpr} ${operation} \`(?i)${model.params[0]}\``;
+    }
     return `${innerExpr} ${operation} \`${model.params[0]}\``;
   };
 }
```
Comment (matyax) [inline public/app/plugins/datasource/loki/querybuilder/operationUtils.ts:268]: I'm inclined to suggest the following alternative implementation:

```suggestion
export function getLineFilterRenderer(operation: string, caseInsensitive?: boolean = false) {
  return function lineFilterRenderer(model: QueryBuilderOperation, def: QueryBuilderOperationDef, innerExpr: string) {
    if (caseInsensitive) {
      return `${innerExpr} ${operation} \`(?i)${model.params[0]}\``;
    }
    return `${innerExpr} ${operation} \`${model.params[0]}\``;
  };
}
```

**Review by matyax — COMMENTED** [review]
Looks good! Left a suggestion.

File: public/app/plugins/datasource/loki/querybuilder/operationUtils.ts (hunk 3 above)
Comment (gwdawson) [reply public/app/plugins/datasource/loki/querybuilder/operationUtils.ts:268]: @matyax much better solution, i'll implement this now.

File: public/app/plugins/datasource/loki/querybuilder/operationUtils.ts [hunk 4]
```diff
@@ -255,8 +255,11 @@ export function addNestedQueryHandler(def: QueryBuilderOperationDef, query: Loki
...
 
-export function getLineFilterRenderer(operation: string) {
+export function getLineFilterRenderer(operation: string, caseInsensitive?: boolean) {
```
Comment (matyax) [inline public/app/plugins/datasource/loki/querybuilder/operationUtils.ts:258]: One more request. Could you please add coverage for this function in `operationUtils.test.ts`. I'd super appreciate that 🙏 

The objective in general is to always back these kind of changes with its corresponding test.

Comment (gwdawson) [reply public/app/plugins/datasource/loki/querybuilder/operationUtils.ts:258]: @matyax ofc, i'll add them now!

**Review by matyax — APPROVED** [review]
Good job!

File: public/app/plugins/datasource/loki/querybuilder/operations.ts [hunk 5]
```diff
@@ -238,6 +238,28 @@ export function getOperationDefinitions(): QueryBuilderOperationDef[] {
       addOperationHandler: addLokiOperation,
       explainHandler: (op) => `Return log lines that contain string \`${op.params[0]}\`.`,
     },
+    {
+      id: LokiOperationId.LineContainsInsensitive,
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/querybuilder/operations.ts:242]: I'd update the order of these operations. As you can see bellow, they are usually grouped (`contains` and then `not contains` under each other). But in this case, they are not. Also, this is probably just personal preference/nit pick, but I'd go with `LineContainsCaseInsensitive` operation name. 

image

File: public/app/plugins/datasource/loki/querybuilder/operations.ts [hunk 6]
```diff
@@ -238,6 +238,28 @@ export function getOperationDefinitions(): QueryBuilderOperationDef[] {
...
+      name: 'Line contains (case insensitive)',
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/querybuilder/operations.ts:243]: I'd name this just `Line contains case insensitive`. I haven't seen any other operation that uses brackets, so I wouldn't introduce new pattern like this.

File: public/app/plugins/datasource/loki/querybuilder/operations.ts [hunk 7]
```diff
@@ -238,6 +238,28 @@ export function getOperationDefinitions(): QueryBuilderOperationDef[] {
...
+          runQueryOnEnter: true,
+        },
+      ],
+      defaultParams: [''],
+      alternativesKey: 'line filter',
+      category: LokiVisualQueryOperationCategory.LineFilters,
+      orderRank: LokiOperationOrder.LineFilters,
+      renderer: getLineFilterRenderer('|~', true),
+      addOperationHandler: addLokiOperation,
+      explainHandler: (op) => `Return log lines that contain string \`${op.params[0]}\`.`,
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/querybuilder/operations.ts:261]: I think that explainHandler here should be `explainHandler: (op) => `Return log lines that match regex \`(?i)${op.params[0]}\`.`,`

**Review by ivanahuckova — COMMENTED** [review]
Nice job! 🚀🙌  I've left some smaller comments bellow.

**Comment by gwdawson** [comment]
> Nice job! 🚀🙌 I've left some smaller comments bellow.

@ivanahuckova thanks for the suggestions, i'll update it now!

File: public/app/plugins/datasource/loki/querybuilder/operationUtils.test.ts [hunk 8]
```diff
@@ -122,3 +124,35 @@ describe('createRangeOperationWithGrouping', () => {
...
+  const MOCK_DEF = undefined as unknown as QueryBuilderOperationDef;
+
+  const MOCK_INNER_EXPR = '{job="grafana"}';
+
+  it('getLineFilterRenderer returns a function', () => {
+    const lineFilterRenderer = getLineFilterRenderer('!~');
+    expect(typeof lineFilterRenderer).toBe('function');
+  });
+
+  it('lineFilterRenderer returns the correct query for (line contains)', () => {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/querybuilder/operationUtils.test.ts:147]:
```suggestion
  it('lineFilterRenderer returns the correct query for line contains', () => {
```

File: public/app/plugins/datasource/loki/querybuilder/operationUtils.test.ts [hunk 9]
```diff
@@ -122,3 +124,35 @@ describe('createRangeOperationWithGrouping', () => {
...
+    const lineFilterRenderer = getLineFilterRenderer('!~');
+    expect(lineFilterRenderer(MOCK_MODEL, MOCK_DEF, MOCK_INNER_EXPR)).toBe('{job="grafana"} !~ `error`');
+  });
+
+  it('lineFilterRenderer returns the correct query for (line contains case insensitive)', () => {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/querybuilder/operationUtils.test.ts:152]:
```suggestion
  it('lineFilterRenderer returns the correct query for line contains case insensitive', () => {
```

**Review by ivanahuckova — APPROVED** [review]
LGTM

**Pull request merged**